File: Python_tutorial/04-StateMgt/bot/state_bot.py
from botbuilder.core import (
    TurnContext,
    ActivityHandler,
    ConversationState,
    UserState
)

from datamodel import ConState,UserProfile,EnumUser

class StateBot(ActivityHandler):
    def __init__(self,constate:ConversationState,userstate:UserState):
        self.constate = constate
        self.userstate = userstate

        self.conprop = self.constate.create_property("constate")
        self.userprop = self.userstate.create_property("userstate")

    async def on_turn(self,turn_context:TurnContext):
        await super().on_turn(turn_context)

        await self.constate.save_changes(turn_context)
        await self.userstate.save_changes(turn_context)

    async def on_message_activity(self,turn_context:TurnContext):
        conmode = await self.conprop.get(turn_context,ConState)
        usermode = await self.userprop.get(turn_context,UserProfile)

        if(conmode.profile == EnumUser.NAME):
            await turn_context.send_activity("Please enter the name")
            conmode.profile = EnumUser.PHONE
        elif(conmode.profile == EnumUser.PHONE):
            usermode.name = turn_context.activity.text
            await turn_context.send_activity("Please enter your phone number")
            conmode.profile = EnumUser.EMAIL
        elif(conmode.profile == EnumUser.EMAIL):
            await turn_context.send_activity("Please enter your email Id")
            usermode.phone = turn_context.activity.text
            conmode.profile = EnumUser.DONE
        elif(conmode.profile == EnumUser.DONE):
            usermode.email = turn_context.activity.text
            info = usermode.name + " " + usermode.phone + "  " + usermode.email
            await turn_context.send_activity(info)
            conmode.profile = EnumUser.NAME

File: csharp_tutorial/BotStateMangement/StateBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;

namespace StateBot
{
    public enum ConTrack
    {
        Name,
        LastName,
        Complete
    }
    public class ConvState
    {
        public ConTrack conTrack = ConTrack.Name;
    }

    public class Userprofile
    {
        public string username { get; set; }
        public string Lastname { get; set; }
    }


    public class StateBotSample : ActivityHandler
    {
        private ConversationState _conversationState;
        private UserState _userstate;
        public StateBotSample(ConversationState constate,UserState userState)
        {
            _conversationState = constate;
            _userstate = userState;
        }
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            foreach (var member in membersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text($"Welcome to Bot"), cancellationToken);
                }
            }
        }

        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var conprop = _conversationState.CreateProperty<ConvState>(nameof(ConvState));
            var conobject = await conprop.GetAsync(turnContext,()=>new ConvState());

            var userProp = _userstate.CreateProperty<Userprofile>(nameof(Userprofile));
            var userobject = await userProp.GetAsync(turnContext, () => new Userprofile());

            //main logic

            switch(conobject.conTrack)
            {
                case ConTrack.Name:
                    await turnContext.SendActivityAsync("Hello , Enter your name");
                    conobject.conTrack = ConTrack.LastName;
                    break;
                case ConTrack.LastName:
                    userobject.username = turnContext.Activity.Text;
                    await turnContext.SendActivityAsync("Hello , Enter your Last Name");
                    conobject.conTrack = ConTrack.Complete;
                    break;
                case ConTrack.Complete:
                    userobject.Lastname = turnContext.Activity.Text;
                    conobject.conTrack = ConTrack.Name;
                    break;
            }


            //if(string.IsNullOrEmpty(userobject.username))
            //{
            //    if(!conobject.username)
            //    {
            //        await turnContext.SendActivityAsync("Hello , Enter your name");
            //        conobject.username = true;
            //    }
            //    else
            //    {
            //        userobject.username = turnContext.Activity.Text;
            //    }
            //}

            await _conversationState.SaveChangesAsync(turnContext,false,cancellationToken);
            await _userstate.SaveChangesAsync(turnContext, false, cancellationToken);

        }
    }
}
